### ISSUE TITLE
OpenAI API Request too large for gpt-3.5-turbo-instruct in organization 

### ISSUE THREAD

**Issue by aseelert** [instructlab/instructlab#2288]
**Describe the bug**
```bash
ilab data generate --pipeline full --gpus 1 --num-cpus 16 --sdg-scale-factor 300 \
--endpoint-url https://api.openai.com/v1 --api-key **projectapikey**  \
--enable-serving-output --output-dir ./outputtest --model gpt-3.5-turbo-instruct \
--batch-size 0 --server-ctx-size 4096 --chunk-word-count 200
```

**To Reproduce**
```python
  File "/home/itzuser/instructlab/venv/lib64/python3.11/site-packages/instructlab/sdg/pipeline.py", line 203, in _generate_single
    raise PipelineBlockError(
instructlab.sdg.pipeline.PipelineBlockError: PipelineBlockError(<class 'instructlab.sdg.llmblock.LLMBlock'>/gen_contexts): Error code: 429 - {'error': {'message': 'Request too large for gpt-3.5-turbo-instruct in organization org-pOOZZmC4ASqo2NMoLudceXRk on tokens per min (TPM): **Limit 90000, Requested 209665**. The input or output tokens must be reduced in order to run successfully. Visit https://platform.openai.com/account/rate-limits to learn more.', 'type': 'tokens', 'param': None, 'code': 'rate_limit_exceeded'}}
```

I tried different models and also other options, but the request is always higher than 90000 (I did not found anything how to reduce this)

**Version**
(venv) [itzuser@itzvsi-550001nu41-hsscbt6c router]$ ilab system info
sys.version: 3.11.2 (main, Jul 18 2024, 00:00:00) [GCC 11.3.1 20221121 (Red Hat 11.3.1-4)]
sys.platform: linux
os.name: posix
platform.release: 5.14.0-284.82.1.el9_2.x86_64
platform.machine: x86_64
platform.node: itzvsi-550001nu41-hsscbt6c
platform.python_version: 3.11.2
os-release.ID: rhel
os-release.VERSION_ID: 9.2
os-release.PRETTY_NAME: Red Hat Enterprise Linux 9.2 (Plow)
instructlab.version: 0.18.4
instructlab-dolomite.version: 0.1.1
instructlab-eval.version: 0.1.2
instructlab-quantize.version: 0.1.0
instructlab-schema.version: 0.3.1
instructlab-sdg.version: 0.2.7
instructlab-training.version: 0.4.2
torch.version: 2.3.1+cu121
torch.backends.cpu.capability: AVX512
torch.version.cuda: 12.1
torch.version.hip: None
torch.cuda.available: True
torch.backends.cuda.is_built: True
torch.backends.mps.is_built: False
torch.backends.mps.is_available: False
torch.cuda.bf16: True
torch.cuda.current.device: 0
torch.cuda.0.name: NVIDIA L4
torch.cuda.0.free: 21.9 GB
torch.cuda.0.total: 22.1 GB
torch.cuda.0.capability: 8.9 (see https://developer.nvidia.com/cuda-gpus#compute)
llama_cpp_python.version: 0.2.90
llama_cpp_python.supports_gpu_offload: True

**Comment by Arturron** [comment]
maybe this will help
https://bit.ly/3TyLBMS
If you don't have the c compliator, install it.(gcc or clang)

**Comment by AbdullahSK03** [comment]
maybe this will help
https://bit.ly/4gwC82r

you may need to install the c compiler

**Comment by AmjedKhaled165** [comment]
try this
https://bit.ly/4gwC82r
If you don't have the c compliator, install it.(gcc or clang)

**Comment by cdoern** [comment]
@bbrowning @aakankshaduggal @khaledsulayman is this a different, yet endpoint-url related, issue than #2287 ?

**Comment by bbrowning** [comment]
Different, yes. It appears we generate payloads too large for gpt-3.5-turbo-instruct, or at least for this user's quota there. However, we also don't test SDG against that endpoint, so it's not a surprising outcome. I guess the question is do we want to keep a list of inference servers we're known to work with somewhere? That list would consist of specific verison ranges of vllm and llama-cpp-python today. Or, do we want to test and support other inference servers for the teacher models?

**Comment by ktam3** [comment]
@booxter - PTAL

**Comment by booxter** [comment]
I don't think we should be in the business of maintaining the list of inference options that we support. Even if we'd have one for today, new entries will pop up tomorrow. We'll spend all day updating the list. (Also existing entries may change their limits, through time or depending on the user profile status or geolocation etc.)

---

What I think we could do here is to sanity check provided endpoints at the very start of any command. The sanity check would poke the endpoint with a request that we believe is enough to validate the endpoint fit for the job; if the request fails, gracefully handle the error, telling the user what happened and what could be their fix (e.g. refer to another endpoint with higher limits.) The CLI would then return code !=0 so that this situation could be spotted in scripts.

(Note: to validate this "feature", we could re-use the stub test-only http server that I proposed to introduce in https://github.com/instructlab/instructlab/issues/2287#issuecomment-2622220042)

**Comment by bbrowning** [comment]
@booxter SDG specifically has some upcoming features that will ONLY work with vllm. They will not work with hosted OpenAI servers, not with llama_cpp_python, etc. If InstructLab wants to open up the list of inference endpoints we support (as opposed to saying we only support vllm for everything and llama_cpp_python for a subset of everything) then we won't be able to do the same in SDG, at least near-term.

**Comment by booxter** [comment]
I see. But even vllm can be served differently, so we could probably give guidance, do some sanity check, but still let the libraries detect and even error on whatever is still missing. (Libraries could also provide their own sanity check endpoints if this is runtime detectable - ilab could trigger these for you.)

**Comment by bbrowning** [comment]
That seems reasonable. Today we attempt to detect vllm vs llama_cpp_python but even that occasionally fails and/or causes issues. In the future, we'll try to use things like vllm guided decoding and may expect only specific configurations of vllm to work (as opposed to all vllm). I don't know how we'll handle any of that currently, but it's definitely a broader discussion to have about what we support and how we detect (or the user configures) which backend they're using.
